### ISSUE TITLE
Option to prevent parentheses from disabling `no-sequences`

### ISSUE THREAD

**Issue by brettz9** [eslint/eslint#5535]
**What version are you using?**

Version on demo.

**What did you do?**

Add the following code:

```js
while ((val = foo(), val < 42));
```

**What happened?**

`no-sequences` is not flagged (because of the extra parentheses).

**What did you expect to happen?**

I would like an option to prevent the parentheses from disabling the rule.

**Comment by nzakas** [comment]
Why do you want such an option, what's your use case?

**Comment by brettz9** [comment]
Allowing it can make for overly complex code (except in for loop initialization or afterthoughts where it is a common, expected convention, and where you already permit the comma operator anyways).

**Comment by michaelficarra** [comment]
@brettz9 I think the question is: why do you feel you need to have the unnecessary grouping parentheses? If you removed them, the rule would report that code.

**Comment by brettz9** [comment]
@michaelficarra My concern was for other contributors to a project who could be prevented from submitting the problematic code.

Since the additional grouping parentheses in the case of sequence expressions are already considered problematic by `no-extra-parens`, one could admittedly add a `no-extra-parens` rule to catch the parenthesized-wrapped sequences too. However, those who wish to prohibit overly complex constructions (with `no-sequences`) may not wish to enable `no-extra-parens`, as they may wish to let their project be lenient in style.

(In the case of `no-cond-assign` and `no-return-assign`, these already have the 'always' option which won't even allow for such error suppression parentheses.)

Maybe my proposal for `no-sequences` could be implemented by adding `always` and `except-parens` options (for parity with the above-mentioned two rules), with `except-parens` being the default for backward compatibility.

Btw, while we're on the topic of parentheses:

1. I think that `no-extra-parens` should document that it does not handle arrow parentheses as does `arrow-parens`. Or alternatively, and for consistency, the `no-extra-parens` rule might be changed to handle them and thereby remove the need for the "as-needed" option with `arrow-parens`.

2. To take the non-problematic range sample from `yoda`:

```js
/*eslint yoda: [2, "never", { "exceptRange": true }]*/

function isReddish(color) {
    return (color.hue < 60 || 300 < color.hue);
}
```

While it is expected that it still reports a `no-extra-parens`, I'd think that `no-extra-parens` should have its own "exceptRange" option, so that someone wishing not to normally allow them could allow them in the case of ranges.

**Comment by nzakas** [comment]
@brettz9 please keep the discussion on topic, we are getting a bit far off now, which makes it hard for others to understand what's being proposed or discussed. 

This issue is about whether or not to add an exception to `no-sequences` such that all sequences are disallowed all of the time. It seems like that is the intent here. 

It's worth noting that you can already do this using the `no-restricted-syntax` rule and configuring it with `SequenceExpression`. Given that, do we need an option added to `no-sequences`?

**Comment by brettz9** [comment]
Sorry, I've edited my previous comment to exclude the parentheses discussion (which I intended to add as a new issue or PR).

I assume the prohibition of `SequenceExpression` by `no-restricted-syntax` will also have the effect of disallowing sequences in for afterthoughts (and indirect evals)--unlike `no-sequences` which explicitly whitelists this particular case.

If that is a correct understanding, then I think @BYK's addition of `ignoreParenthesized` option is still useful since some (including myself) may wish to disallow all sequences including double-parentheses-escaped ones, with the exception of sequences in for afterthoughts (where it is common, including for side effects, and not as likely to be confusing as within, say, a condition). The same goes for @BYK's exception made for indirect eval by `allowIndirectEval`.

**Comment by BYK** [comment]
@nzakas - after seeing your `no-restricted-syntax` suggestion, I'm a bit skeptical about the necessity of this change.

**Comment by BYK** [comment]
But in all honesty, in that light, I think `no-sequence` itself may be completely obsolete.

**Comment by brettz9** [comment]
@BYK, what about people who want all sequences except for indirect eval (such as you've added) and/or for loop afterthoughts (which `no-sequences` specifically permits)?

**Comment by BYK** [comment]
@brettz9 I think that's an edge case. That said we can probably cover it with `no-eval`.

**Comment by brettz9** [comment]
What about allowing for sequences in loop afterthoughts? Those are pretty common...

**Comment by nzakas** [comment]
@brettz9 Our rules are designed to cover the most common cases and will definitely not be able to be configurable down to the level you're asking about. If you really want this many edge cases, then your best bet is to create a custom rule.

**Comment by brettz9** [comment]
But it already is configurable down to that degree in a sense. You explicitly make an exception in `no-sequences` for for afterthoughts, but if you get rid of the `no-sequences` rule (a possibility mentioned by @BYK), then one is no longer able to do this out of the box.


My suggestions may be pretty granular, btw (though I think I'm done now, besides a few doc suggestions), but I think all of them are either cases envisaged by others (like the afterthought and indirect eval mentioned here), are seeking harmony between rules, or are seeking control corresponding directly to a feature of the language (like the directives issue), none of which do I see as really edge cases.

Of course, if you want to get rid of the rule or the new code by BYK, fine, I'm very happy that I can still adapt the `no-sequences` code with options added from @BYK regardless of your decision.

**Comment by nzakas** [comment]
We aren't getting rid of `no-sequences`, it's been around for a while and I think the exception case is useful. I think the best part forward here is for you to create a custom rule that does exactly what you want or else use `no-restricted-syntax` to block all sequences.
